pkg/chain/ethereum/ethutil/rate_limiter_test.go

```diff
@@ -0,0 +1,246 @@
+	"time"
+)
+
+const requestDuration = 250 * time.Millisecond
+
+func TestRateLimiter(t *testing.T) {
+	requestsPerSecondLimit := 200
+	concurrencyLimit := 50
+	acquirePermitTimeout := time.Minute
+	requests := 1000
+
+	backend := &mockBackend{make([]string, 0), sync.Mutex{}}
+
+	rateLimitingBackend := WrapRateLimiting(
+		backend,
+		&RateLimiterConfig{
+			RequestsPerSecondLimit: requestsPerSecondLimit,
+			ConcurrencyLimit:       concurrencyLimit,
+			AcquirePermitTimeout:   acquirePermitTimeout,
+		},
+	)
+
+	wg := sync.WaitGroup{}
+	wg.Add(requests)
+
+	startSignal := make(chan struct{})
+
+	for i := 0; i < requests; i++ {
+		go func() {
+			<-startSignal
+
+			err := rateLimitingBackend.SendTransaction(context.Background(), nil)
+			if err != nil {
+				t.Errorf("unexpected error: [%v]", err)
+			}
+
+			wg.Done()
+		}()
+	}
+
+	startTime := time.Now()
+	close(startSignal)
+
+	wg.Wait()
+
+	duration := time.Now().Sub(startTime)
+	averageRequestsPerSecond := float64(requests) / duration.Seconds()
+
+	if averageRequestsPerSecond > float64(requestsPerSecondLimit) {
+		t.Errorf(
+			"average requests per second exceeded the limit\n"+
+				"limit:  [%v]\n"+
+				"actual: [%v]",
+			requestsPerSecondLimit,
+			averageRequestsPerSecond,
+		)
+	}
+
+	maxConcurrency := 0
+	temporaryConcurrency := 0
+	for _, event := range backend.events {
+		if event == "start" {
+			temporaryConcurrency++
+		}
+
+		if event == "end" {
+			temporaryConcurrency--
+		}
+
+		if temporaryConcurrency > maxConcurrency {
+			maxConcurrency = temporaryConcurrency
+		}
+	}
+
+	if maxConcurrency > concurrencyLimit {
+		t.Errorf(
+			"max concurrency exceeded the limit\n"+
+				"limit:  [%v]\n"+
+				"actual: [%v]",
+			concurrencyLimit,
+			maxConcurrency,
+		)
+	}
+
+	fmt.Printf(
+		"actual average requests per second: [%v]\n"+
+			"actual maximum concurrency: [%v]\n",
+		averageRequestsPerSecond,
+		maxConcurrency,
+	)
+}
+
+func TestRateLimiter_AcquirePermitTimout(t *testing.T) {
+	requestsPerSecondLimit := 1
+	concurrencyLimit := 1
+	acquirePermitTimeout := 10 * time.Millisecond
+	requests := 3
+
+	backend := &mockBackend{make([]string, 0), sync.Mutex{}}
+
+	rateLimitingBackend := WrapRateLimiting(
+		backend,
+		&RateLimiterConfig{
+			RequestsPerSecondLimit: requestsPerSecondLimit,
+			ConcurrencyLimit:       concurrencyLimit,
+			AcquirePermitTimeout:   acquirePermitTimeout,
+		},
+	)
+
+	wg := sync.WaitGroup{}
+	wg.Add(requests)
+
+	startSignal := make(chan struct{})
+	errorChannel := make(chan error, requests)
+	errors := make([]error, 0)
+
+	go func() {
+		for e := range errorChannel {
+			errors = append(errors, e)
+		}
+	}()
+
+	for i := 0; i < requests; i++ {
+		go func() {
+			<-startSignal
+
+			err := rateLimitingBackend.SendTransaction(context.Background(), nil)
+			if err != nil {
+				errorChannel <- err
+			}
+
+			wg.Done()
+		}()
+	}
+
+	close(startSignal)
+
+	wg.Wait()
+
+	if len(errors) == 0 {
+		t.Fatalf("at least one timeout error should be present")
```

Comment:
At least one or at least two given that `requestDuration = 250 * time.Millisecond` and `acquirePermitTimeout := 10 * time.Millisecond`?

Reply:
This actually doesn't matter. We only need to make sure the timeout works.